### PR TITLE
Ensure all tests clean up after themselves #383

### PR DESCRIPTION
**Pull request by edesdan**
Hi Guys, I did a little bit of work for the #383 issue. So basically now after you run the functional tests all the created docker images and docker containers are automatically removed. I also added two new methods to start/remove the docker registry container that was causing 3 tests to fail. Let me know what you think.

### CODE REVIEW

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy [hunk 1]
```diff
@@ -34,6 +45,20 @@ abstract class AbstractFunctionalTest extends Specification {
 
     String dockerServerUrl
 
+    @Shared
+    DockerClientConfig config = DefaultDockerClientConfig.createDefaultConfigBuilder()
+        .withDockerHost("unix:///var/run/docker.sock")
```
Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: Now tests won't run on Windows. It would be better to use `TestConfiguration` here.

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy [hunk 2]
```diff
@@ -63,6 +88,105 @@ abstract class AbstractFunctionalTest extends Specification {
...
+                    .exec()
+            } catch (NotFoundException ex) {
+                println("Exception while removing image: ${ex.message} ")
+            }
+        }
+
+        println("Docker images removed.")
+    }
+
+    protected void startDockerRegistryContainer() {
```
Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: I would change it to `startDockerRegistryIfNeeded`

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy [hunk 3]
```diff
@@ -63,6 +88,105 @@ abstract class AbstractFunctionalTest extends Specification {
...
+
+        def containers = dockerClient.listContainersCmd().exec()
+        println("Active dockerContainersBeforeTests: ${containers}")
```
Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: This prints lots of junk in console, maybe use only container ids?

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy [hunk 4]
```diff
@@ -63,6 +88,105 @@ abstract class AbstractFunctionalTest extends Specification {
...
+
+        boolean isRegistryAlreadyRunning
+
+        isRegistryAlreadyRunning = containers.find {
+            return it?.names?.contains("/registry")
```
Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: `return` can be omitted.

Comment (orzeh) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: in most cases this should work, however if I start registry container with custom name it will fail. But for now it's :ok_hand:

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy (hunk 1 above)
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: This is why I suggested to use the AbstractDockerTask and grab the client from there. It removes the need to do this as well as define a docker-java version in yet another place.

Comment (edesdan) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: yeah, I agree it is not the best, I'll will review this part.

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy (hunk 2 above)
Comment (edesdan) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: ehm not sure I'll think about it...

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy (hunk 3 above)
Comment (edesdan) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: sure, I will remove the prints

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy (hunk 4 above)
Comment (edesdan) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: Yeah you are right but for now I will go with that..

File: gradle/functional-test.gradle [hunk 5]
```diff
@@ -30,6 +30,7 @@ check.dependsOn functionalTest
 
 dependencies {
     functionalTestCompile 'commons-lang:commons-lang:2.6'
+    functionalTestCompile 'com.github.docker-java:docker-java:3.0.12'
```
Comment (cdancy) [inline gradle/functional-test.gradle]: As I noted below: if we use the AbstractDockerTask we don't need to rely upon bringing this onto the `functionalTestCompile` classpath.

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy [hunk 6]
```diff
@@ -63,6 +88,105 @@ abstract class AbstractFunctionalTest extends Specification {
...
+                .withForce(true)
+                .exec()
+        } catch (NotFoundException ex) {
+            println("Exception while removing container: ${ex.message} ")
+        }
+    }
+
+
+
+
```
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: Lots of extra newlines added. Lets remove these.

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy [hunk 7]
```diff
@@ -115,7 +239,7 @@ abstract class AbstractFunctionalTest extends Specification {
         "gradle/${generateRandomUUID()}"
     }
 
-    protected String createUniqueContainerName() {
+    protected final String createUniqueContainerName() {
```
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: Why this change?

**Comment by cdancy** [comment]
My only complaint is that I think we may be over thinking this. From my POV I don't think we need to worry about which images/containers were around before the start of testing so long as we have a standard naming convention (which we do to some extent and I'm open to change it) and can simply iterate, once testing has completed, through all of the remaining images and containers and remove those that match our naming convention.

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy (hunk 3 above)
Comment (cdancy) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: If we do end up keeping this piece of code I would change it to `logger.info` as I still think it's useful just not something we would care to see for every run.

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy (hunk 7 above)
Comment (edesdan) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: I guess it got in the mix during my tests... I will remove it.

**Comment by edesdan** [comment]
@cdancy I see your point but I think this is no over thinking. I think it is more safe doing it in this way and 'self' maintainable (basically you don't need to rely on any specific logic for naming containers and images). I tried going in the other direction first and I saw the drawbacks. Anyway that's my POV. I'll try to update the PR ASAP. Thanks.

**Comment by cdancy** [comment]
@edesdan fair enough. Feel free to continue going down your road I would just make sure to use the abstract class so that we don't have to include another dependency.

File: src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy (hunk 1 above)
Comment (edesdan) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: @cdancy can you point me to the code where I can grab the client? I cannot find any AbstractDockerTask class. Thanks.

Comment (edesdan) [inline src/functTest/groovy/com/bmuschko/gradle/docker/AbstractFunctionalTest.groovy]: @orzeh I see how I can use TestConfiguration class to get the docker config. I guess @cdancy suggestion is to avoid bringing the client dependency but I'm still not finding any reference where to see how to grab a docker-client.

**Comment by edesdan** [comment]
I did another commit to address the minor comments. Regarding no adding a new dependency I need help from you guys, as I sad in the previous comments I don't know how to address the issue. If you just can past some code snippet I'm happy to do the change. Thanks.

**Comment by cdancy** [comment]
@edesdan [have a look here](https://github.com/bmuschko/gradle-docker-plugin/blob/84939895a69e48c2f277d56868be855a67a9c994/src/functTest/groovy/com/bmuschko/gradle/docker/AbstractDockerRemoteApiTaskFunctionalTest.groovy#L21). It would be much easier to implement a custom "cleanup task" of some sort and just purge them all this way IMO.

**Comment by edesdan** [comment]
@cdancy What you mean by 'much easier'? I'm not getting your point. I had a look at the link you sent me but still I'm not sure what I should do. If it is not a big issue for you I will take this work I have done and merge the PR : it works good to me and I think I have addresses all minor issues raised after the first commit. Feel free to merge it and change it, or feel free to reject it. I don't think I can spend more time on this, sorry about that.

Anyway I think this should be a priority, it could stop people to collaborate to the project. Honestly when I first executed the tests I had a look at my docker images/containers it was a mess: waste of space and memory.

**Comment by cdancy** [comment]
@edesdan I'll try to put something together this week but the main idea is:

If you use a custom `AbstractDockerRemoteApiTask` you don't have to add a functional test dependency on `docker-java` which is yet another thing we will have to maintain. Instead you get passed an instance of `dockerClient` to do whatever you want with:

```
            // using pseudo-code here to get the point across
            class CleanupTask extends AbstractDockerRemoteApiTask {
                @Override
                void runRemoteCommand(dockerClient) {
                    def listOfGeneratedImage = dockerClient.getListOfImagesSomehow()
                    dockerClient.removeImages(listOfGeneratedImage)

                    def listOfGeneratedContainers = dockerClient.getListOfContainersSomehow()
                    dockerClient.removeContainers(listOfGeneratedContainers)
                }
            }
```

**Comment by cdancy** [comment]
Closing for now as this PR has been open almost 2 years and I think there are more complete solutions we can take.
